src/Components/Endpoints/src/Builder/RazorComponentEndpointDataSource.cs

```diff
@@ -77,17 +86,57 @@ private void UpdateEndpoints()
     {
         var endpoints = new List<Endpoint>();
         var context = _builder.Build();
+        if (_renderModes.RenderModes == null)
+        {
+            _renderModes.RenderModes = new();
+            _renderModes.RenderModes.AddRange(context.GetDeclaredRenderModesByDiscoveredComponents());
+        }
+
+        var renderModes = _renderModes.RenderModes;
+
         foreach (var definition in context.Pages)
         {
             _factory.AddEndpoints(endpoints, typeof(TRootComponent), definition, _conventions, _finallyConventions);
         }
 
+        var found = false;
+        for (var i = 0; i < renderModes.Count; i++)
+        {
+            var renderMode = renderModes[i];
+            foreach (var provider in _renderModeEndpointProviders)
+            {
+                if (provider.Supports(renderMode))
+                {
+                    found = true;
+                    RenderModeEndpointProvider.AddEndpoints(
+                        endpoints,
+                        typeof(TRootComponent),
+                        provider.GetEndpointBuilders(renderMode, _applicationBuilder.New()),
+                        renderMode,
+                        _conventions,
+                        _finallyConventions);
+                }
+            }
+
+            if (!found)
+            {
+                throw new InvalidOperationException($"Unable to find a provider for the render mode: {renderMode.GetType().FullName}. This generally" +
+                    $"means that a call to 'AddWebAssemblyComponents' or 'AddServerComponents' is missing. " +
+                    $"Alternatively call 'UseWebAssemblyRendering', 'UseServerRendering' or 'UseStaticRendering' " +
+                    $"to configure the desired rendering mode.");
+            }
+        }
```

Comment:
Will there be unit tests for this?

Reply:
https://github.com/dotnet/aspnetcore/pull/48283/files#diff-40cc1e09f579892477787e39faa6f3e76be7a88828528c943e8614fa44093f2b